### PR TITLE
Extend secrets volumes with path control

### PR DESCRIPTION
**Pull request by ingvagabund**
As per [1] this PR extends secrets mapped into volume with:
- key-to-path mapping the same way as is for configmap. E.g.

```
{
 "apiVersion": "v1",
 "kind": "Pod",
  "metadata": {
    "name": "mypod",
    "namespace": "default"
  },
  "spec": {
    "containers": [{
      "name": "mypod",
      "image": "redis",
      "volumeMounts": [{
        "name": "foo",
        "mountPath": "/etc/foo",
        "readOnly": true
      }]
    }],
    "volumes": [{
      "name": "foo",
      "secret": {
        "secretName": "mysecret",
        "items": [{
          "key": "username",
          "path": "my-username"
        }]
      }
    }]
  }
}
```

Here the `spec.volumes[0].secret.items` added changing original target `/etc/foo/username` to `/etc/foo/my-username`.
- secondly, refactoring `pkg/volumes/secrets/secrets.go` volume plugin to use `AtomicWritter` to project a secret into file.

[1] https://github.com/kubernetes/kubernetes/blob/master/docs/design/configmap.md#changes-to-secret

### CODE REVIEW

**Comment by pmorie** [comment]
#24744

**Comment by pmorie** [comment]
@bgrant0607 for API review

File: pkg/volume/secret/secret.go [hunk 1]
```diff
@@ -169,36 +150,68 @@ func (b *secretVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
...
+	}
 
 	return nil
 }
 
+func makePayload(mappings []api.KeyToPath, secret *api.Secret) (map[string][]byte, error) {
+	payload := make(map[string][]byte, len(secret.Data))
+
+	if len(mappings) == 0 {
+		for name, data := range secret.Data {
```
Comment (pmorie) [inline pkg/volume/secret/secret.go]: I frequently wish that go allowed interface specifications to contain fields or had a mixin mechanism that allow type-unspecific use of fields with a specific type signature.

**Comment by pmorie** [comment]
@ingvagabund looks great so far, will you add an e2e as well?

**Comment by pmorie** [comment]
@ingvagabund also, doc changes to kubernetes.github.io in a follow-up PR in that project.

**Comment by ingvagabund** [comment]
Docs updated https://github.com/kubernetes/kubernetes.github.io/pull/478

**Comment by ingvagabund** [comment]
File permissions are changing as well:
Originally, the secret file had 0444 permissions. With introduction of key-to-path mapping, the secret root directory can contain subdirectories as well. Thus, the rule is like this:

1. the secret volume is created as EmptyDir volume with 777 permissions for the target directory

   ```bash
   stat -c "%a %n" /etc/foo
   1777 /etc/foo
   ```

2. the AtomicWriter creates timestamp directory with 755 permissions

   ```
   stat -c "%a %n" /etc/foo/..5985_11_05_12_35_30.806564262
   755 /etc/foo/..5985_11_05_12_35_30.806564262
   ```

3. the AtomicWriter creates for each secret corresponding file with 644 permissions, basedir(file) with 777 permissions
   ```
   stat -c "%a %n" /etc/foo/..5985_11_05_12_35_30.806564262/path
755 /etc/foo/..5985_11_05_12_35_30.806564262/path
   stat -c "%a %n" /etc/foo/..5985_11_05_12_35_30.806564262/path/my-username
644 /etc/foo/..5985_11_05_12_35_30.806564262/path/my-username
   ```

4. ``/etc/foo/..data`` now points to the timestamp directory, the previous one is deleted

Additional data:

```bash
ls -ld /etc/foo/
drwxrwxrwt. 4 root root 100 May 11 12:35 /etc/foo/
ls -ld /etc/foo/path
drwxr-xr-x. 2 root root 60 May 11 12:35 /etc/foo/path
ls -ld /etc/foo/path/my-username
lrwxrwxrwx. 1 root root 26 May 11 12:35 /etc/foo/path/my-username -> ../..data/path/my-username
ls -ld /etc/foo/..data/path/my-username
-rw-r--r--. 1 root root 8 May 11 12:35 /etc/foo/..data/path/my-username
```

``path`` is symlinked to ``..5985_11_05_12_35_30.806564262/path`` as expected:

```bash
ls -ld /etc/foo/..5985_11_05_12_35_30.806564262/path/
drwxr-xr-x. 2 root root 60 May 11 12:35 /etc/foo/..5985_11_05_12_35_30.806564262/path/
ls -ld /etc/foo/..5985_11_05_12_35_30.806564262/path/my-username
-rw-r--r--. 1 root root 8 May 11 12:35 /etc/foo/..5985_11_05_12_35_30.806564262/path/my-username
```

Still confused as both ``/etc/foo/path/my-username`` and ``/etc/foo/..data/path/my-username`` should have the same permission:

```
stat -c "%a %n" /etc/foo/path/my-username
777 /etc/foo/path/my-username
stat -c "%a %n" /etc/foo/..data/path/my-username
644 /etc/foo/..data/path/my-username
```

**Comment by bgrant0607** [comment]
I'm fine with this change. It's consistent with ConfigMapVolumeSource.

**Comment by ingvagabund** [comment]
Updated e2e tests for ``secret volume``. New version of mounttest image is required. This PR depends on https://github.com/kubernetes/kubernetes/pull/25800.

**Comment by ingvagabund** [comment]
Wrt. to the permissions:

* as the default root mask is ``0022``, the permissions 0777 are correctly changed to 0755.
* for the symlink case, the target is correctly set to 0644 as expected, for the symlink itself the 777 permissions is quite normal [1]

[1] http://stackoverflow.com/questions/7487793/symbolic-link-not-inheriting-permissions

**Comment by ncdc** [comment]
@k8s-bot e2e test this issue: #25800

**Comment by pmorie** [comment]
@ingvagabund now that #25800 has merged, do you need to update this PR?

**Comment by ingvagabund** [comment]
@pmorie the PR can be merge as it is

**Comment by pmorie** [comment]
@ingvagabund thanks a lot for the fix

**Pull request merged**